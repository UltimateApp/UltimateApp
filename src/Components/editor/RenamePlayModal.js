import React, { useEffect, useRef } from 'react';
import { StyleSheet, View, Text, TextInput, InteractionManager, ScrollView } from 'react-native';
import { connect } from 'react-redux';
import { Formik } from 'formik';
import * as Yup from 'yup';

import I18n from '../../utils/i18n';
import { showSuccess } from '../../utils/flashMessage';
import { renamePlay } from '../../Store/Actions/playAction';
import Modal from '../shared/Modal';
import Button from '../shared/Button';

export const RenamePlayModal = (props) => {
  const inputRef = useRef();

  useEffect(() => {
    InteractionManager.runAfterInteractions(() => {
      inputRef.current.focus();
    });
  }, []);

  return (
    <View>
      <Modal
        title={props.currentPlay.title || I18n.t('playEditorPage.untitledPlay')}
        visible
        onClose={() => {
          props.close();
        }}
      >
        {/* keyboardShouldPersistTaps enables to keep the keyboard visible.*/
        /* Otherwise the first press on the validation button will only close the keyboard and not trigger onPress */}
        <ScrollView keyboardShouldPersistTaps="always" directionalLockEnabled style={{ flexGrow: 0 }}>
          <Formik
            initialValues={{ name: props.currentPlay.title }}
            validationSchema={Yup.object({
              name: Yup.string()
                .trim()
                .required(I18n.t('editor.renamePlayModal.empty'))
                .notOneOf(
                  props.customPlays.map((play) => play.title),
                  I18n.t('editor.renamePlayModal.alreadyExists'),
                ),
            })}
            onSubmit={(values) => {
              props.renamePlay(props.currentPlay.uuid, values.name);
              props.currentPlay.title = values.name;
              props.onRename();
              showSuccess(I18n.t('editor.renamePlayModal.renameSuccess'));
              props.close();
            }}
          >
            {({ handleSubmit, handleChange, errors, values, touched, isValid }) => (
              <View style={styles.form}>
                <TextInput
                  style={styles.input}
                  placeholder={I18n.t('editor.renamePlayModal.placeholder')}
                  onChangeText={handleChange('name')}
                  value={values.name}
                  ref={inputRef}
                />
                {touched.name && errors.name && <Text style={styles.error}>{errors.name}</Text>}
                <Button
                  onPress={handleSubmit}
                  text={I18n.t('editor.renamePlayModal.cta')}
                  style={styles.cta}
                  disabled={!isValid}
                />
              </View>
            )}
          </Formik>
        </ScrollView>
      </Modal>
    </View>
  );
};

const mapStateToProps = (state) => {
  return {
    customPlays: state.customPlays,
  };
};

const mapDispatchToProps = { renamePlay };

export default connect(mapStateToProps, mapDispatchToProps)(RenamePlayModal);

const styles = StyleSheet.create({
  form: {
    alignItems: 'center',
  },
  input: {
    flexBasis: '80%',
    marginHorizontal: 5,
  },
  cta: {
    marginTop: 10,
    width: 120,
  },
  error: {
    fontStyle: 'italic',
    color: 'red',
  },
});
