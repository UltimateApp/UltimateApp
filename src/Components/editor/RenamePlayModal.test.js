import React from 'react';
import { render, waitFor, fireEvent, act } from '@testing-library/react-native';
import { Provider as PaperProvider } from 'react-native-paper';

import animationSquare from '../../Fixtures/Animation/AnimationSquare';

import { RenamePlayModal } from './RenamePlayModal';

describe('<RenamePlayModal />', () => {
  afterEach(() => jest.clearAllMocks());

  const onRename = jest.fn();
  const close = jest.fn();
  const currentPlay = { play: animationSquare, title: 'Square' };
  const customPlays = [currentPlay];
  const renamePlay = jest.fn();

  it('renders correctly', async () => {
    const { toJSON } = await waitFor(() =>
      render(
        <PaperProvider>
          <RenamePlayModal
            currentPlay={currentPlay}
            onRename={onRename}
            close={close}
            customPlays={customPlays}
            renamePlay={renamePlay}
          />
        </PaperProvider>,
      ),
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('does nothing when using Back', async () => {
    const { getByPlaceholderText, getByText } = await waitFor(() =>
      render(
        <PaperProvider>
          <RenamePlayModal
            currentPlay={currentPlay}
            onRename={onRename}
            close={close}
            customPlays={customPlays}
            renamePlay={renamePlay}
          />
        </PaperProvider>,
      ),
    );

    // Fill input
    fireEvent(getByPlaceholderText('Click here to enter the new name'), 'onChangeText', 'New Title');

    await act(async () => await fireEvent.press(getByText('Back')));

    expect(onRename).not.toBeCalled();
    expect(renamePlay).not.toBeCalled();
    expect(currentPlay.title).toEqual('Square');
  });

  it('renames the play on validation', async () => {
    const { getByPlaceholderText, getByText } = await waitFor(() =>
      render(
        <PaperProvider>
          <RenamePlayModal
            currentPlay={currentPlay}
            onRename={onRename}
            close={close}
            customPlays={customPlays}
            renamePlay={renamePlay}
          />
        </PaperProvider>,
      ),
    );

    // Fill input
    fireEvent(getByPlaceholderText('Click here to enter the new name'), 'onChangeText', 'New Title');

    await act(async () => await fireEvent.press(getByText('Apply')));

    expect(onRename).toBeCalled();
    expect(renamePlay).toBeCalled();
    expect(currentPlay.title).toEqual('New Title');
  });
});
