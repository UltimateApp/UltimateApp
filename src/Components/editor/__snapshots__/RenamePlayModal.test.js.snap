// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RenamePlayModal /> renders correctly 1`] = `
<View>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    useNativeDriver={true}
    visible={true}
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(0,0,0,0.5)",
          "bottom": 0,
          "left": 0,
          "opacity": 1,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderRadius": 10,
            "elevation": 10,
            "maxHeight": "80%",
            "maxWidth": "80%",
            "padding": 20,
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.25,
            "shadowRadius": 3.84,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 20,
              "fontWeight": "bold",
              "marginBottom": 20,
              "textAlign": "center",
            }
          }
        >
          Square
        </Text>
        <RCTScrollView
          directionalLockEnabled={true}
          keyboardShouldPersistTaps="always"
          style={
            Object {
              "flexGrow": 0,
            }
          }
        >
          <View>
            <View
              style={
                Object {
                  "alignItems": "center",
                }
              }
            >
              <TextInput
                allowFontScaling={true}
                onChangeText={[Function]}
                placeholder="Click here to enter the new name"
                rejectResponderTermination={true}
                style={
                  Object {
                    "flexBasis": "80%",
                    "marginHorizontal": 5,
                  }
                }
                underlineColorAndroid="transparent"
                value="Square"
              />
              <View
                accessible={true}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#2a9e91",
                    "borderRadius": 5,
                    "flexDirection": "row",
                    "height": 50,
                    "justifyContent": "center",
                    "marginTop": 10,
                    "opacity": 1,
                    "paddingHorizontal": 20,
                    "paddingVertical": 10,
                    "width": 120,
                  }
                }
                testID="button"
              >
                <Text
                  style={
                    Object {
                      "color": "#fff",
                      "fontSize": 16,
                      "textAlign": "center",
                    }
                  }
                >
                  Apply
                </Text>
              </View>
            </View>
          </View>
        </RCTScrollView>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#fff",
              "borderColor": "#2a9e91",
              "borderRadius": 5,
              "borderWidth": 1,
              "flexDirection": "row",
              "height": 50,
              "justifyContent": "center",
              "marginTop": 20,
              "opacity": 1,
              "paddingHorizontal": 20,
              "paddingVertical": 10,
              "width": 120,
            }
          }
          testID="button"
        >
          <Text
            style={
              Object {
                "color": "#2a9e91",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            Back
          </Text>
        </View>
      </View>
    </View>
  </Modal>
</View>
`;
